<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>prolfqua&amp;colon; A Comprehensive R-package for Proteomics Differential Expression Analysis • prolfquabenchmark</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="prolfqua&amp;colon; A Comprehensive R-package for Proteomics Differential Expression Analysis">
<meta property="og:description" content="prolfquabenchmark">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">prolfquabenchmark</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BenchmarkMSFraggerProteinIonStar.html">Benchmarking MSFragger output using Ionstar Dataset</a>
    </li>
    <li>
      <a href="../articles/Benchmark_MSStats.html">Benchmarking MSstats using the Ionstar Dataset</a>
    </li>
    <li>
      <a href="../articles/Benchmark_SaintExpress.html">Benchmarking SaintExpress</a>
    </li>
    <li>
      <a href="../articles/Benchmark_proDA_fromMQlfq.html">Benchmarking the proDA package using the Ionstar Dataset MQ LFQ intensities</a>
    </li>
    <li>
      <a href="../articles/Benchmark_proDA_medpolish.html">Benchmarking the proDA package using the Ionstar Dataset starting from peptides</a>
    </li>
    <li>
      <a href="../articles/BenchmarkingIonstarData.html">Benchmarking normalization, aggregation and models using the Ionstar Dataset</a>
    </li>
    <li>
      <a href="../articles/paper.html">prolfqua&amp;colon; A Comprehensive R-package for Proteomics Differential Expression Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wolski/proLFQuaBenchmark/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>prolfqua: A Comprehensive R-package for
Proteomics Differential Expression Analysis</h1>
                        <h4 data-toc-skip class="author">Witold E.
Wolski <span class="math inline">\(^{\dagger\ddagger}\)</span><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 data-toc-skip class="author">Paolo
Nanni<span class="math inline">\(^{\ast}\)</span>
</h4>
                        <h4 data-toc-skip class="author">Jonas
Grossmann<span class="math inline">\(^{\ast\ddagger}\)</span>
</h4>
                        <h4 data-toc-skip class="author">Maria
d’Errico<span class="math inline">\(^{\ast\ddagger}\)</span>
</h4>
                        <h4 data-toc-skip class="author">Ralph
Schlapbach<span class="math inline">\(^{\ast}\)</span>
</h4>
                        <h4 data-toc-skip class="author">Christian
Panse<span class="math inline">\(^{\ast\ddagger}\)</span>
</h4>
            
            <h4 data-toc-skip class="date">2022-06-09 15:35:12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/wolski/proLFQuaBenchmark/blob/HEAD/../vignettes/paper.Rmd" class="external-link"><code>../vignettes/paper.Rmd</code></a></small>
      <div class="hidden name"><code>paper.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Mass spectrometry is widely used for quantitative proteomics
      studies, relative protein quantification, and differential
      expression analysis of proteins. Nevertheless, there is a need for
      a flexible and easy-to-use application programming interface in R
      that transparently supports a variety of well principled
      statistical procedures. The <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> package
      can model simple experimental designs with a single explanatory
      variable and complex experiments with multiple factors and
      hypothesis testing. It integrates essential steps of the mass
      spectrometry-based differential expression analysis workflow:
      quality control, data normalization, protein aggregation,
      statistical modeling, hypothesis testing, and sample size
      estimation. The application programmer interface strives to be
      clear, predictable, discoverable, and consistent to make
      proteomics data analysis easy and exciting. Furthermore, the
      package implements benchmark functionality that can help to
      compare data acquisition, data preprocessing, or data modeling
      methods using a gold standard dataset. Finally, we show that the
      implemented methods allow sensitive and specific differential
      expression analysis. The <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> R
      package is available on GitHub <a href="https://github.com/fgcz/prolfqua" class="external-link uri">https://github.com/fgcz/prolfqua</a>, distributed
      under the MIT licence, and runs on all platforms supported by the
      <a href="https://www.r-project.org/" class="external-link">R free software environment
      for statistical computing and graphics</a>.</p>
    </div>
    
<div style="page-break-after: always;"></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Proteins carry out most crucial functions and give structure to
living cells. Hence, measuring changes in protein abundance is the
subject of active research <span class="citation">(Vidova and Spacil
2017)</span>. Bottom-up mass spectrometric methods, where proteins are
specifically and reproducibly digested into protein fragments -
peptides, are employed to identify and quantify proteins in complex
samples containing hundreds to thousand of proteins <span class="citation">(Bubis et al. 2017; Veiga Leprevost et al.
2020)</span>. The peptides are first separated by their chemical and
physical properties using liquid chromatography (LC). Afterward, they
are ionised, weighed, identified, and quantified using mass
spectrometric techniques, e.g., electro-spray-ionization mass
spectrometry (ESI-MS). Finally, peptide identification is achieved by
fragmenting and matching the measured fragment masses to theoretical
masses computed from known peptide sequences <span class="citation">(Eng
et al. 2015; Yu, Li, and Yu 2016; Kong et al. 2017)</span>. For
quantification, intact peptide ions <span class="citation">(Yu et al.
2020; Cox and Mann 2008)</span> or products of peptide ion fragmentation
<span class="citation">(Röst et al. 2014; Demichev et al. 2020)</span>
are counted and aggregated to obtain peptide abundances. Finally, the
identified and quantified peptides are assigned to proteins based on
protein sequence information.</p>
<p>Proteomics quantification experiments enable monitoring relative
abundances of thousands of proteins in biological samples. Most studies
use parallel-group designs, where one or many treatment groups are
compared to the control group <span class="citation">(Leeuw et al. 2022;
Laubscher et al. 2021)</span>. More recently, more complex experimental
designs with an increasing number of samples are studied, e.g.,
factorial designs and longitudinal studies (time series), sometimes with
repeated measurements on the same subject <span class="citation">(Tan et
al. 2022; Meier-Abt et al. 2021)</span>. The data can be modeled using
linear fixed-, mixed-, or random-effects models <span class="citation">(Bates et al. 2015)</span>. Based on these models,
tests can be applied to examine whether specific factors and factor
interactions are significant, e.g., it can be tested if differences in
protein abundance between groups are statistically significant.</p>
<p>An important aspect of mass spectrometric data are missing peptide
and protein quantifications. <span class="citation">Rubin (1976)</span>
classified missing data problems into three categories: missing
completely at random (MCAR), missing at random (MAR), and missing not at
random (MNAR). For instance, in data-dependent acquisition (DDA) MS,
only a limited number of MS1 signals are selected for fragmentation and
identified. Peptide quantification algorithms transfer identification
information between MS1 features in different samples by aligning the
data using retention time and mass information, thus reducing the amount
of missing data. Nevertheless, highly abundant proteins can suppress the
detection of other proteins in a sample, a MAR mechanism. Furthermore, a
weak correlation between the number of missing measurements in a group
and the abundance of the quantified protein can be observed, which is
caused by the limit of detection (LOD), a MNAR mechanism <span class="citation">(McGurk et al. 2020)</span>.</p>
<p>Several data analysis packages exist to model mass spectrometry
protein quantification experiments, e.g., <em><a href="https://bioconductor.org/packages/3.14/limma" class="external-link">limma</a></em> <span class="citation">(Ritchie et al. 2015)</span>, <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
<span class="citation">(Choi et al. 2014)</span>, <em><a href="https://bioconductor.org/packages/3.14/PECA" class="external-link">PECA</a></em> <span class="citation">(Suomi and Elo 2017)</span>, <em><a href="https://bioconductor.org/packages/3.14/msqrob2" class="external-link">msqrob2</a></em>
<span class="citation">(Goeminne, Gevaert, and Clement 2016)</span> or
<em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em> <span class="citation">(Ahlmann-Eltze and Anders 2020)</span>, to mention
some, all implemented in R. Each of them has some unique features, for
example, <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
infers the model and generates the model formula from the sample
annotation, allowing users with limited statistical knowledge to perform
differential expression analysis (DEA). At the same time, <em><a href="https://bioconductor.org/packages/3.14/limma" class="external-link">limma</a></em>
allows to specify a design matrix using a linear model formula and
implements the empirical Bayes variance shrinkage method. The package
<em><a href="https://bioconductor.org/packages/3.14/PECA" class="external-link">PECA</a></em>,
performs a roll-up of peptide level differences and peptide level <span class="math inline">\(p\)</span>-value estimates, obtained from <em><a href="https://bioconductor.org/packages/3.14/limma" class="external-link">limma</a></em> or
<em><a href="https://bioconductor.org/packages/3.14/PECA" class="external-link">PECA</a></em>,
to protein level estimates. Furthermore, <em><a href="https://bioconductor.org/packages/3.14/msqrob2" class="external-link">msqrob2</a></em>
implements two models: robust linear models fitted to protein
intensities and robust ridge regression fitted to peptide intensities,
and combines them with empirical Bayes variance shrinkage. The <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>
package implements a linear probabilistic dropout model to account for
missing data without imputation. <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
handles missing data by feature filtering and using imputation. Other
means of modelling missing observations are the Hurdle model discussed
by <span class="citation">Goeminne et al. (2020)</span>, while the R
package <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>
models missingness using probabilistic dropout models <span class="citation">(Ahlmann-Eltze and Anders 2020)</span>.</p>
<p>When analyzing parallel-group designs using a single explanatory
variable, and contrasting groups, we can use all R packages; but, we can
use only some of them if we want to model factorial designs or repeated
measurements. Table @ref(tab:tableOverview) gives an overview of the
models and features supported by these packages. We see that packages
such as <em><a href="https://bioconductor.org/packages/3.14/limma" class="external-link">limma</a></em> and
<em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em> allow
us to fit a comprehensive variety of models and test various hypotheses;
however, in-depth knowledge of design matrix specification using the R
formula interface is required <span class="citation">(Law et al.
2020)</span>.</p>
<table class="table">
<caption>Rows - R package, Columns - types of models supported: pd -
parallel design, rm - repeated measurements, fd - factorial design, int
- interactions among factors, mem - mixed effect models, eb - empirical
Bayes, md - missing data modelling. Y - yes, * - repeated measurements
are modeled using a fixed effect. ? - the hurdle model was published but
is not available in the msqrob2 package.</caption>
<thead><tr class="header">
<th align="left"></th>
<th align="left">pd</th>
<th align="left">rm</th>
<th align="left">mem</th>
<th align="left">fd</th>
<th align="left">int</th>
<th align="left">eb</th>
<th align="left">md</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MSstats</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">ROPECA</td>
<td align="left">Y</td>
<td align="left">Y*</td>
<td align="left">NA</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">limma</td>
<td align="left">Y</td>
<td align="left">Y*</td>
<td align="left">NA</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">MSqRob2_rlm</td>
<td align="left">Y</td>
<td align="left">Y*</td>
<td align="left">NA</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">?</td>
</tr>
<tr class="odd">
<td align="left">proDA</td>
<td align="left">Y</td>
<td align="left">Y*</td>
<td align="left">NA</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
</tr>
<tr class="even">
<td align="left">prolfqua</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
<td align="left">Y</td>
</tr>
</tbody>
</table>
<p>When developing the R package <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> we were
inspired by the R package <em><a href="https://CRAN.R-project.org/package=caret" class="external-link">caret</a></em> <span class="citation">(Kuhn 2008)</span>, which enables us to call various
machine learning (ML) methods, and makes selecting the best ML algorithm
for your problem easy. We aimed for a similar R package for the
differential expression analysis of proteomics data. However, after
examining the R packages for modeling proteomics quantification data, we
observed that model specification, input, and output formats differ
widely. At the same time, they have in common that: they fit linear
models either to peptide or protein intensities, determine differences
among groups, and afterward apply empirical Bayes variance shrinkage.
Therefore, the revised goal was to provide a modular object-oriented
(OO) design, with R linear models as a core, where we can add features
such as <span class="math inline">\(p\)</span>-value aggregation, e.g.,
<em><a href="https://bioconductor.org/packages/3.14/PECA" class="external-link">PECA</a></em>,
variance shrinkage, or modeling of missing observations.</p>
<p>Furthermore, the functionality of <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> also includes
methods specific to proteomics data. For example, we implemented
strategies to estimate protein intensities from peptide intensities: top
N <span class="citation">(Grossmann et al. 2010)</span>, Tukey’s median
polish <span class="citation">(Tukey et al. 1977)</span>, robust linear
models <span class="citation">(Goeminne et al. 2020)</span>.
Furthermore, peptide or protein abundances can be scaled and
transformed, using robust scaling, <em>quantile</em> normalization or
<em><a href="https://bioconductor.org/packages/3.14/vsn" class="external-link">vsn</a></em> to
remove systematic differences among samples and heteroscedasticity. In
this respect, <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> can be
compared with other R packages such as <em><a href="https://bioconductor.org/packages/3.14/DEP" class="external-link">DEP</a></em> <span class="citation">(Zhang et al. 2018)</span> or <em><a href="https://bioconductor.org/packages/3.14/POMA" class="external-link">POMA</a></em> <span class="citation">(Castellano-Escuder, Andrés-Lacueva, and Sánchez-Pla
2021)</span> which support the entire differential expression analysis
pipeline.</p>
<p>We also implemented functionality and use the Ionstar <span class="citation">(Shen et al. 2018)</span> dataset to benchmark the
modeling methods implemented within <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> and to compare
our results with those of <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
and <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>.
Since group sizes are relatively small, typically with four or five
subjects per group, the high power of the tests is a relevant criterion
to assess the modeling method. The quantified proteins can be ranked
using the estimated fold-change, <span class="math inline">\(t\)</span>-statistics, or scaled <span class="math inline">\(p\)</span>-value, and afterward subjected to gene
set enrichment (GSEA) or over-representation analysis <span class="citation">(Subramanian et al. 2005)</span> to determine up or
down-regulated groups of proteins. Furthermore, the statistical model
must provide an unbiased estimate of the false discovery rate (FDR) to
manage expectations when selecting protein lists for follow-up
experiments. We will use the partial area under the receiver operator
curve (ROC) to assess the power of the tests and compare the FDR with
the false discovery proportion (FDP).</p>
<div style="page-break-after: always;"></div>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<div class="section level3">
<h3 id="implementation">Implementation<a class="anchor" aria-label="anchor" href="#implementation"></a>
</h3>
<p>We store all the data needed for analysis in a tidy table, i.e.,
every column is a variable, every row is an observation, and every cell
is a single value <span class="citation">(Wickham 2014)</span>. Using an
R6 <span class="citation">(Chang 2020)</span> configuration object
(Figure @ref(fig:LFQData)), we specify what variable is in which column,
making it easy to integrate new inputs in <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> if provided as
a tidy table. For example, to visualize tidy Spectronaut<span class="citation">(Bruderer et al. 2015)</span>, DiaNN<span class="citation">(Demichev et al. 2020)</span>, or Skyline<span class="citation">(MacLean et al. 2010)</span> outputs, or data in <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em><span class="citation">(Choi et al. 2014)</span> format, only a few lines of
code are needed to update the <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>
<code>AnalysisTableConfiguration</code> configuration. The configuration
encapsulates the differences in column names among the various input
formats, and enables the use of the methods without additional
arguments. We show an example code for creating an MSFragger<span class="citation">(Yu et al. 2020)</span> configuration in the Appendix.
For popular software like MaxQuant<span class="citation">(Cox and Mann
2008)</span>, or MSFragger, which stores the same variable, e.g.,
intensity, in multiple columns, one for each sample, we implemented
methods that transform the data into tidy tables. Relying on the tidy
data table enables us to interface with many data manipulation,
visualization, and modeling methods, implemented in base R <span class="citation">(R Core Team 2021)</span> and the tidyverse <span class="citation">(Wickham et al. 2019)</span>, easily. We use R6 classes
to structure the functionality of the package (see Figure
@ref(fig:LFQData) and Figure @ref(fig:ContrastUML)). R6 classes are well
supported by command-line completion features in RStudio <span class="citation">(RStudio Team 2022)</span>, and help to implement
argument free functions (Figure @ref(fig:tabCompletion).</p>
<p>(ref:LFQData) Class Diagram of classes representing the proteomics
data. The LFQData class encapsulates the quantitative proteomics data
stored in a tidy table. An instance of the AnalysisTableConfiguration
class specifies a mapping of columns in the tidy table. The
LFQDataPlotter class and other classes decorate the LFQData class with
additional functionality. Similarly, the <code>LFQDataStats</code> and
<code>LFQDataSummary</code> reference the <code>LFQData</code> class and
group methods for variance and sample size estimation or summarizing
peptide and protein counts.</p>
<div class="figure">
<img src="LFQData_UML_V2.png" alt="(ref:LFQData)" width="90%"><p class="caption">
(ref:LFQData)
</p>
</div>
<p>In addition, we implement features specific to high throughput
experiments, such as the experimental Bayes variance and <span class="math inline">\(p\)</span>-value moderation, which utilizes the
parallel structure of the protein measurements and the analysis <span class="citation">(Ritchie et al. 2015)</span>. We also compute
probabilities of differential protein regulation based on peptide level
models <span class="citation">(Suomi and Elo 2017)</span>. We used R6
classes to encapsulate the statistical modelling functionality in <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> (see Figure
@ref(fig:ContrastUML)). We specify a contrast interface
(<code>ContrastsInterface</code>). Several implementations allow to
perform differential expression analysis given linear or mixed effect
models (<code>Contrasts</code>), variance shrinkage
(<code>ContrastsModerated</code>), or to impute contrasts in cases when
observations are missing for an entire group of samples
(<code>ContrastsSimpleImpute</code>). Further implementations of the
interface encapsulate and integrate differential expression analysis
results of external tools such as <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em> or of
SAINTexpress <span class="citation">(Teo et al. 2014)</span> used to
analyze data from protein interaction studies.</p>
<p>(ref:ContrastUML) The UML (unified modeling language) diagram of
modeling and contrast related classes. Different strategies, e.g.,
<em>lm</em>, <em>lmer</em>, and <em>glm</em> (Table
@ref(tab:prolfquaModels)), can fit various types of models. The model
builder method fits the statistical model given the data and a strategy.
The obtained model can then be used for the analysis of variance or to
estimate contrasts. All classes estimating contrasts implement the
<em>Contrasts</em> interface. Results of external tools, e.g.,
SAINTexpress, can be adapted by implementing the Contrasts
Interface.</p>
<div class="figure">
<img src="ContrastClassesUML.png" alt="(ref:ContrastUML)" width="90%"><p class="caption">
(ref:ContrastUML)
</p>
</div>
</div>
<div class="section level3">
<h3 id="dataset-for-benchmarking">Dataset for benchmarking<a class="anchor" aria-label="anchor" href="#dataset-for-benchmarking"></a>
</h3>
<p>To evaluate the performance of differential expression analysis, we
use the IonStar benchmark dataset<span class="citation">(Shen et al.
2018)</span>, available from the Proteomics Identifications Database
(PRIDE) identifier PXD003881. <span class="math inline">\(DH5\alpha\)</span> <em>E. coli</em> lysate was
spiked in human pancreatic cancer cells (Panc-1) lysate at five
different levels: <span class="math inline">\(3\%\)</span>, <span class="math inline">\(4.5\%\)</span>, <span class="math inline">\(6\%\)</span>, <span class="math inline">\(7.5\%\)</span> and <span class="math inline">\(9\%\)</span> <em>E. coli</em>. We annotated these
dilutions from smallest to largest with the letters <em>a</em>,
<em>b</em>, <em>c</em>, <em>d</em>, <em>e</em>. By comparing the various
dilutions, we obtain different effect sizes, e.g., when comparing
dilution <em>e</em> (9%) against dilution <em>d</em> (7.5%), the
expected difference is <span class="math inline">\(1.2\)</span> for E.
coli proteins and <span class="math inline">\(1\)</span> for human
proteins. There are four technical replicates for each dilution, and
hence 20 raw files in total. We processed the raw data using MaxQuant
<span class="citation">(Cox and Mann 2008)</span> version Version
1.6.10.43, with MaxQuant default settings. The text files generated by
MaxQuant are available in the <em>prolfquadata</em> R package <span class="citation">(Wolski 2021)</span>. To compare the performance of the
various methods implemented in <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> we use only
the contrasts resulting in minor differences <span class="math inline">\(\Delta = (1.20, 1.25, 1.30, 1.50)\)</span>,
because for bigger differences all models perform similarly.</p>
</div>
<div class="section level3">
<h3 id="data-preprocessing-for-model-comparison">Data preprocessing for model comparison<a class="anchor" aria-label="anchor" href="#data-preprocessing-for-model-comparison"></a>
</h3>
<p>The peptide abundances (from the MaxQuant <em>peptide.txt</em> file)
were <span class="math inline">\(log_2\)</span> transformed and
subsequently scaled, where median and the mean average deviation (mad)
were obtained from the human proteins only. We removed <em>one hit
wonders</em>, i.e., proteins with a single peptide assignment. Protein
abundances are inferred from the peptide intensities using Tukey’s
median polish. Finally, we fitted the fixed effect linear models, the
dropout model <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em> to
protein abundances, the mixed effect linear model, and the <em><a href="https://bioconductor.org/packages/3.14/PECA" class="external-link">PECA</a></em> model
to peptide intensities.</p>
</div>
<div class="section level3">
<h3 id="benchmark-metrics">Benchmark metrics<a class="anchor" aria-label="anchor" href="#benchmark-metrics"></a>
</h3>
<p>The IonStar dataset contains <em>H. sapiens</em> proteins with
constant concentrations and <em>E. coli</em> proteins with varying
concentrations. We know that for <em>H. sapiens</em> proteins, the
difference <span class="math inline">\(\beta\)</span> between two
dilutions should be <span class="math inline">\(\beta = 0\)</span>,
while for <em>E. coli</em> proteins, we know that the difference between
dilutions should be <span class="math inline">\(\beta \ne
0\)</span>.</p>
<p>We can use various statistics to examine the alternative hypothesis
<span class="math inline">\(\beta \ne 0\)</span>: the contrast estimate,
i.e. the <span class="math inline">\(\log_2\)</span> fold-change <span class="math inline">\(\beta\)</span>, the <span class="math inline">\(t\)</span>-statistic <span class="math inline">\(\frac{\beta}{\sqrt{var(\beta)}}\)</span>, or the
<span class="math inline">\(p\)</span>-value and moderated <span class="math inline">\(p\)</span>-value. For each statistic and each
value of the statistics we then compute a confusion matrix (see Table
@ref(tab:knitrConfusionMatrix)). From the confusion matrix we obtain
measures such as true positive rate (<span class="math inline">\(TPR\)</span>), false positive rate (<span class="math inline">\(FPR\)</span>), or false discovery proportion
(<span class="math inline">\(FDP\)</span>) which are given by:</p>
<table class="table">
<caption>Confusion matrix, TP - true positive, FP - false positive, FN -
false negative, TN - true negatives, P - all positive cases (all <span class="math inline">\(E.~coli\)</span> proteins), N - all negative cases
(all <span class="math inline">\(H.~sapiens\)</span> proteins), m - all
proteins.</caption>
<thead><tr class="header">
<th align="left">Prediction  Truth</th>
<th align="left">E.coli</th>
<th align="left">H.sapiens</th>
<th align="left">Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">beta != 0</td>
<td align="left">TP</td>
<td align="left">FP</td>
<td align="left">R</td>
</tr>
<tr class="even">
<td align="left">beta == 0</td>
<td align="left">FN</td>
<td align="left">TN</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Total</td>
<td align="left">P</td>
<td align="left">N</td>
<td align="left">m</td>
</tr>
</tbody>
</table>
<p>with</p>
<p><span class="math display">\[\begin{eqnarray}
TPR &amp;= \frac{TP}{TP+FN} = \frac{TP}{P}\\
FPR &amp;= \frac{FP}{FP+TN} = \frac{FP}{N}\\
FDP &amp;= \frac{FP}{TP + FP} = \frac{FP}{R}
\end{eqnarray}\]</span></p>
<p>By plotting the <span class="math inline">\(TPR\)</span> versus the
<span class="math inline">\(FPR\)</span> we obtain the receiver operator
characteristic curve (ROC curve). The area under the curve (AUC) or
partial areas under the curve (pAUC), at various values of the <span class="math inline">\(FPR\)</span>, are measures of performance derived
from the ROC curve. By using these measures we can compare the
performances of the statistics produced by the various methods
examined.</p>
<p>In order to compute the confusion matrices based on the <span class="math inline">\(p\)</span>-value we first need to rescale it (see
Equation @ref(eq:scalepvalue)). Thus, the <span class="math inline">\(p\)</span>-value is a function of the <span class="math inline">\(t\)</span>-statistics and the degrees of
freedom.</p>
</div>
<div class="section level3">
<h3 id="modelling">Modelling<a class="anchor" aria-label="anchor" href="#modelling"></a>
</h3>
<div class="section level4">
<h4 id="robust-scaling-of-the-data">Robust scaling of the data<a class="anchor" aria-label="anchor" href="#robust-scaling-of-the-data"></a>
</h4>
<p><span class="citation">Välikangas, Suomi, and Elo (2016)</span>
discuss and benchmark various methods of peptide or protein intensity
normalization, such as variance stabilizing normalization <span class="citation">(Huber et al. 2002)</span> or quantile normalization
<span class="citation">(Bolstad et al. 2003)</span>. In this work, we
use a robust version of the z-score, where instead of the mean <span class="math inline">\(\bar{x}\)</span> we use the median, and instead of
the standard deviation <span class="math inline">\(\tilde{S}\)</span> we
use the median absolute deviation (mad):</p>
<p><span class="math display">\[\begin{eqnarray}
z = \frac{x - \tilde{x}}{\tilde{S}}
\end{eqnarray}\]</span></p>
<p>Because we need to estimate the differences among groups on the
original scale, we must multiply the <span class="math inline">\(z\)</span>-score by the average variance of all
the <span class="math inline">\(n\)</span> samples in the
experiment.</p>
<p><span class="math display">\[\begin{eqnarray}
z' = z \cdot \frac{1}{n}\sum_{i=1}^n S_i
\end{eqnarray}\]</span></p>
<p>To apply this transformation, we need to estimate two parameters per
sample, therefore it works for experiments with thousands of proteins
and experiments where only a few hundred proteins per sample are
measured. For the Ionstar dataset, we used the intensities of <em>H.
sapiens</em> proteins, whose concentrations do not change, to determine
<span class="math inline">\(\bar{x}\)</span> and <span class="math inline">\(S\)</span> and then applied it to all the
intensities (including <em>E. coli</em>) in the sample.</p>
</div>
<div class="section level4">
<h4 id="estimating-differences-between-groups">Estimating differences between groups<a class="anchor" aria-label="anchor" href="#estimating-differences-between-groups"></a>
</h4>
<p>Given a linear model <span class="math inline">\(y = \beta
X\)</span>, we can compute the difference <span class="math inline">\(\beta_{c}\)</span> between two groups by a linear
combination <span class="math inline">\(c\)</span> of linear model
parameters <span class="math inline">\(\beta\)</span>, where <span class="math inline">\(c\)</span> is a column vector with as many
elements as there are coefficients <span class="math inline">\(\beta\)</span> in the linear model. If <span class="math inline">\(c\)</span> has <span class="math inline">\(0\)</span> for one or more of its rows, then the
corresponding coefficient in <span class="math inline">\(\beta\)</span>
is not involved in determining the contrast <span class="citation">(Irizarry and Love 2018)</span>.</p>
<p>The difference estimate <span class="math inline">\(\beta_c\)</span>,
is given by the dot product:</p>
<p><span class="math display">\[\begin{eqnarray}
\hat{\beta_{c}} &amp;=&amp; c^T \beta
\end{eqnarray}\]</span></p>
<p>and the variance of <span class="math inline">\(\beta_c\)</span>
by:</p>
<p><span class="math display">\[\begin{eqnarray}
\textrm{var} (\hat\beta_c) &amp;=&amp; \sqrt{ c^T \sigma^2 (X^T X)^{-1}
c }
\end{eqnarray}\]</span></p>
<p>with <span class="math inline">\(X\)</span> being the design matrix.
The degrees of freedom for the contrast are equal to the residual
degrees of freedom of the linear model. For estimating contrasts from
mixed effects models we used the function <code>contest</code>
implemented in the R package <em><a href="https://CRAN.R-project.org/package=lmerTest" class="external-link">lmerTest</a></em> and
used the Satterthwaite <span class="citation">(Kuznetsova, Brockhoff,
and Christensen 2017)</span> method to estimate the denominator degrees
of freedom. These methods are available in the class
<code>Contrast</code> (see Figure @ref(fig:ContrastUML))</p>
</div>
<div class="section level4">
<h4 id="determining-linear-parameter-combinations-for-treatment-comparison">Determining linear parameter combinations for treatment
comparison<a class="anchor" aria-label="anchor" href="#determining-linear-parameter-combinations-for-treatment-comparison"></a>
</h4>
<p>The package <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> provides a
function to determine the vector of <span class="math inline">\(parameter\)</span> weights <span class="math inline">\(c\)</span>, from a linear models and a contrast
specification string.</p>
<p>The linear model below explain the observed protein abundances using
the explanatory variables <code>factor_1</code> and
<code>factor_2</code> and the interaction among them
<code>factor_1:factor_2</code>,</p>
<pre><code><span class="co">## Intensity ~ factor_1 + factor_2 + factor_1:factor_2</span></code></pre>
<p>then the contrasts among <code>group_1</code> and
<code>group_2</code> defined by <code>factor_1</code> can be specified
using the string below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"contrast_name"</span> <span class="op">=</span> <span class="st">"factor_1group_1 - factor_1group_2"</span><span class="op">)</span></code></pre></div>
<p>Furthermore, contrasts for subgroups can be specified using the code
below,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"contrast_name"</span> <span class="op">=</span>
    <span class="st">"`factor_1group_1:factor_2group_a` - `factor_1group_1:factor_2group_b`"</span><span class="op">)</span></code></pre></div>
<p>where <code>factor_x</code> is the name of the explanatory variable,
and <code>group_x</code> are group labels.</p>
<p>The following code shows an example where we specify two contrast:
the first to compare Cells of type CMP/MEP with cells of type HSC, and
the second to compare therapy NO with therapy NU for the celltype
CMP/MEP <span class="citation">(Meier-Abt et al. 2021)</span>. Finally,
we compute the array of weights defining the contrast, used to multiply
the model coefficient <span class="math inline">\(\beta\)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Contrasts</span> <span class="op">&lt;-</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CMP/MEPvsHSC"</span> <span class="op">=</span> <span class="st">"`CelltypeCMP/MEP` - `CelltypeHSC`"</span>,
   <span class="st">"NOvsHU"</span> <span class="op">=</span> <span class="st">"`class_therapyc.NO:CelltypeCMP/MEP` - `class_therapyp.HU:CelltypeCMP/MEP`"</span><span class="op">)</span>

<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/prolfqua_data.html" class="external-link">prolfqua_data</a></span><span class="op">(</span><span class="st">'data_basicModel_p1807'</span><span class="op">)</span>
<span class="va">linfct</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/linfct_from_model.html" class="external-link">linfct_from_model</a></span><span class="op">(</span><span class="va">m</span>,as_list <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/linfct_matrix_contrasts.html" class="external-link">linfct_matrix_contrasts</a></span><span class="op">(</span><span class="va">linfct</span>, <span class="va">Contrasts</span> <span class="op">)</span></code></pre></div>
<pre><code><span class="co">##              (Intercept) class_therapyc.NO class_therapyp.NO</span>
<span class="co">## CMP/MEPvsHSC           0                 0                 0</span>
<span class="co">## NOvsHU                 0                 1                 0</span>
<span class="co">##              CelltypeHSC</span>
<span class="co">## CMP/MEPvsHSC          -1</span>
<span class="co">## NOvsHU                 0</span></code></pre>
</div>
<div class="section level4">
<h4 id="contrast-estimation-in-presence-of-missing-data-">Contrast estimation in presence of missing data.<a class="anchor" aria-label="anchor" href="#contrast-estimation-in-presence-of-missing-data-"></a>
</h4>
<p>Missing observations lead to different group sizes, which results in
unbalanced designs. Linear and mixed effect models can handle unbalanced
designs, as long as one observation per group is available, they will
produce unbiased estimates, and no imputation is needed.</p>
<p>If there is no observation in a group, we assume that the protein is
unobserved because the protein abundance is below the limit of detection
(LOD) of the MS instrument. In this case, we will impute the mean using
the protein abundance at the detection limit <span class="math inline">\(A_{LOD}\)</span>. We estimate the abundance at the
detection limit using the abundances of the proteins observed only in
one sample of a group of samples. Then, we compute the median of the
distribution and use it as the group mean if a protein is absent in a
treatment group.</p>
<p>When computing differences <span class="math inline">\(\Delta\)</span> among two groups <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span>, we will use the group mean <span class="math inline">\(\bar{a}\)</span> or <span class="math inline">\(\bar{b}\)</span> estimated from the data, or if no
observations are present in a group, we use <span class="math inline">\(A_{LOD}\)</span>, e.g., If there are no
observations in group <span class="math inline">\(b\)</span> then :</p>
<p><span class="math display">\[
\Delta =
\begin{cases}
\bar{a} - A_{LOD} &amp; \text{if} ~~ \bar{a} &gt; A_{LOD}\\
0  &amp;\text{if} ~~ \bar{a} &lt; A_{LOD}.
\end{cases}
\]</span></p>
<p>To estimate the variance, we assume that the variance of the protein
is constant in all the groups and use the pooled variance based on all
data:</p>
<p><span class="math display">\[\begin{equation}
s_p^2=\frac{\sum_{i=1}^k (n_i - 1)s_i^2}{\sum_{i=1}^k (n_i - 1)}
\end{equation}\]</span></p>
<p>with <span class="math inline">\(n_i\)</span> the number of
observations, and <span class="math inline">\(s_i\)</span> the standard
deviation in each group. The standard deviation for the <span class="math inline">\(t\)</span>-statistics is then given by:</p>
<p><span class="math display">\[\begin{equation}
s = \sqrt{\frac{2n_g s_p^2}{n}},
\end{equation}\]</span></p>
<p>Where, <span class="math inline">\(n_g\)</span> is the number of
groups and <span class="math inline">\(n\)</span> is the number of
observations. If variance can not be estimated for a protein, we will
use the median pooled variance of all the proteins.</p>
<p>This methods are implemented in the class
<code>ContrastSimpleImpute</code> (see Figure
@ref(fig:ContrastUML)).</p>
</div>
<div class="section level4">
<h4 id="p-value-moderation">
<span class="math inline">\(p\)</span>-value moderation<a class="anchor" aria-label="anchor" href="#p-value-moderation"></a>
</h4>
<p>From the linear and the mixed effect models, we can obtain the
residual standard deviation <span class="math inline">\(\sigma\)</span>,
and degrees of freedom <span class="math inline">\(df\)</span>. <span class="citation">Smyth (2004)</span> discuss how, to use the <span class="math inline">\(\sigma\)</span> and <span class="math inline">\(df\)</span> of all models to estimate a prior
<span class="math inline">\(\sigma\)</span> and prior <span class="math inline">\(df\)</span>, and posterior <span class="math inline">\(\tilde \sigma\)</span>. These can be used to
moderate the <span class="math inline">\(t\)</span>-statistics by:</p>
<p><span class="math display">\[\begin{equation}
\tilde{t}_{pj} = \frac{t_{pj} s_p}{\tilde{s}_p}.
\end{equation}\]</span></p>
<p>We implemented this method in the class
<code>ContrastModerated</code> (Figure @ref(fig:ContrastUML)).</p>
</div>
<div class="section level4">
<h4 id="summarizing-peptide-level-differences-and-p-values-on-protein-level">Summarizing peptide level differences and p-values on protein
level<a class="anchor" aria-label="anchor" href="#summarizing-peptide-level-differences-and-p-values-on-protein-level"></a>
</h4>
<p>To summarize peptide level models to protein models, we apply the
method suggested by <span class="citation">Suomi and Elo (2017)</span>
that use the median scaled <span class="math inline">\(p\)</span>-value
of the peptide models and the cumulative distribution function of the
Beta distribution (CDF) to determine a regulation probability of the
protein.</p>
<p>To obtain the <span class="math inline">\(\tilde{p}\)</span> of a
protein, we first rescaled the peptide <span class="math inline">\(p\)</span>-values by taking the sign of the
fold-change <span class="math inline">\(\hat \beta\)</span> into
account, i.e.:</p>
<p><span class="math display">\[\begin{equation}
p_{s} =
  \begin{cases}
1-p, &amp; \textrm{if}~ \hat{\beta} &gt; 0\\
p-1, &amp; \textrm{otherwise}
\end{cases}
(\#eq:scalepvalue)
\end{equation}\]</span></p>
<p>Afterwards, the median scaled <span class="math inline">\(p\)</span>-value <span class="math inline">\(\tilde{p}_s\)</span> is determined and using the
transformation below, transformed back onto the original scale:</p>
<p><span class="math display">\[\begin{equation}
\tilde{p} = 1 - |\tilde{p}_{s}|
\end{equation}\]</span></p>
<p>Because we use the median, with the i-th order statistic <span class="math inline">\(i = \frac{n}{2} + 0.5\)</span>, we parametrize the
CDF of the Beta distribution with <span class="math inline">\(\gamma = i
= \frac{n}{2} + 0.5\)</span> and <span class="math inline">\(\delta = n
- i + 1 = n - (\frac{n}{2} + 0.5) + 1 = \frac{n}{2} + 0.5 =
\gamma\)</span>. We implemented this method in the class
<code>ContrastROPECA</code> (Figure @ref(fig:ContrastUML)).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="results-and-discussion-">Results and Discussion.<a class="anchor" aria-label="anchor" href="#results-and-discussion-"></a>
</h2>
<div class="section level3">
<h3 id="example-analysis-workflow">Example analysis workflow<a class="anchor" aria-label="anchor" href="#example-analysis-workflow"></a>
</h3>
<p>The code snippets in this section demonstrate how a differential
expression analysis workflow can be implemented using the <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> R package. To
speed up the computation of these examples, we use a random subset of
the Ionstar dataset containing <span class="math inline">\(163\)</span>
proteins and <span class="math inline">\(1258\)</span> peptides. Peptide
abundances are <span class="math inline">\(\log_2\)</span> transformed
and robust z-score scaled using the method <code>robscale</code>. Using
the <code>LFQDataPlotter</code> class, we show the distribution of the
normalized peptide abundances in Figure @ref(fig:prepro) Panel A.
Afterwards, protein intensities are estimated from peptide intensities
using Tukey’s median polish. Figure @ref(fig:prepro) Panel B shows the
peptide intensities and the estimated protein intensities. Next, we
compute the standard deviation of all the proteins in each group and
display their distribution using violin plots (Panel C). Finally, we
create a boxplot (Panel D) showing the abundance of one protein.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## load peptide level data</span>
<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/prolfqua_data.html" class="external-link">prolfqua_data</a></span><span class="op">(</span><span class="st">'data_ionstar'</span><span class="op">)</span><span class="op">$</span><span class="fu">filtered</span><span class="op">(</span><span class="op">)</span>

<span class="co">## create R6 obejct</span>
<span class="va">lfqd</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/LFQData.html" class="external-link">LFQData</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">d</span><span class="op">$</span><span class="va">data</span>, <span class="va">d</span><span class="op">$</span><span class="va">config</span><span class="op">)</span> 

<span class="co">##  transform intensities</span>
<span class="va">t</span> <span class="op">&lt;-</span> <span class="va">lfqd</span><span class="op">$</span><span class="fu">get_Transformer</span><span class="op">(</span><span class="op">)</span>

<span class="va">lfqd</span> <span class="op">&lt;-</span>  <span class="va">t</span><span class="op">$</span><span class="fu">log2</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">robscale</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">lfq</span>
<span class="va">lfqd</span><span class="op">$</span><span class="fu">rename_response</span><span class="op">(</span><span class="st">"peptide_abundance"</span><span class="op">)</span>

<span class="co">## infer protein intensities from peptide intensity</span>
<span class="va">agr</span> <span class="op">&lt;-</span> <span class="va">lfqd</span><span class="op">$</span><span class="fu">get_Aggregator</span><span class="op">(</span><span class="op">)</span>
<span class="va">lfqp</span> <span class="op">&lt;-</span> <span class="va">agr</span><span class="op">$</span><span class="fu">medpolish</span><span class="op">(</span><span class="op">)</span>
<span class="va">lfqp</span><span class="op">$</span><span class="fu">rename_response</span><span class="op">(</span><span class="st">"protein_abundance"</span><span class="op">)</span>

<span class="co">## plot panels A-D</span>
<span class="va">pl</span> <span class="op">&lt;-</span> <span class="va">lfqd</span><span class="op">$</span><span class="fu">get_Plotter</span><span class="op">(</span><span class="op">)</span>
<span class="va">panelA</span> <span class="op">&lt;-</span> <span class="va">pl</span><span class="op">$</span><span class="fu">intensity_distribution_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>
<span class="va">panelB</span> <span class="op">&lt;-</span> <span class="va">agr</span><span class="op">$</span><span class="fu">plot</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span>
<span class="va">panelC</span> <span class="op">&lt;-</span> <span class="va">lfqp</span><span class="op">$</span><span class="fu">get_Stats</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">violin</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span>
<span class="va">pl</span> <span class="op">&lt;-</span> <span class="va">lfqp</span><span class="op">$</span><span class="fu">get_Plotter</span><span class="op">(</span><span class="op">)</span>
<span class="va">panelD</span> <span class="op">&lt;-</span> <span class="va">pl</span><span class="op">$</span><span class="fu">boxplots</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">boxplot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">panelA</span>, <span class="va">panelB</span>, <span class="va">panelC</span>, <span class="va">panelD</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="paper_files/figure-html/prepro-1.png" alt="(ref:scaling)" width="90%"><p class="caption">
(ref:scaling)
</p>
</div>
<p>(ref:scaling) Panel A - Peptide intensity distributions for <span class="math inline">\(20\)</span> samples. For each sample a line with a
different colour is shown. Panel B - Peptide intensities for protein
<em>5NTD</em> are shown using a line of different colour, and the
protein intensity estimate is shown using a fat black line, Panel C -
distribution of standard deviations of all proteins in each dilution
group (<span class="math inline">\(a\)</span>, <span class="math inline">\(b\)</span>, <span class="math inline">\(c\)</span>, <span class="math inline">\(d\)</span>, <span class="math inline">\(e\)</span>) and overall (all), Panel D -
Distribution of protein intensities for protein <em>5NTD</em>.</p>
<p>The following code example illustrates how we compute differences
among groups. First, the linear model and the differences are specified.
Afterward, the model is fitted to the data using the
<code>build_model</code> function. Next, we estimate the contrasts
either from the linear model using the <code>Contrasts</code> class or
directly from the data using the <code>ContrastsSimpleImpute</code>
class. Afterward, we apply <span class="math inline">\(t\)</span>-statistic moderation using the
<code>ContrastModerated</code> class. Finally, the
<code>addContrastResults</code> function merges both sets of contrast
estimates, preferring the one obtained from the linear model if both are
available. Then we create the plots shown in Figure
@ref(fig:exampleContrasts). Panel A shows the distribution of the <span class="math inline">\(p\)</span>-values, while Panel B shows the volcano
plot for each comparison.</p>
<p>(ref:exampleContrasts) Panel A - Histogram showing the distribution
of p-values for <span class="math inline">\(163\)</span> proteins. Panel
B - Volcano plot showing <span class="math inline">\(-\log_{10}\)</span>
transformed FDR as function of the difference between groups for <span class="math inline">\(163\)</span> proteins.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># specify differences among groups</span>
<span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  <span class="st">"dilution_(9/7.5)_1.2"</span> <span class="op">=</span>   <span class="st">"dilution.e - dilution.d"</span>,
  <span class="st">"dilution_(7.5/6)_1.25"</span> <span class="op">=</span>   <span class="st">"dilution.d - dilution.c"</span>
<span class="op">)</span>
<span class="co">## fit model</span>
<span class="va">lmmodel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">lfqp</span><span class="op">$</span><span class="fu">intensity_column</span><span class="op">(</span><span class="op">)</span>,<span class="st">" ~ dilution."</span><span class="op">)</span>
<span class="va">modelFunction</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/strategy.html" class="external-link">strategy_lm</a></span><span class="op">(</span> <span class="va">lmmodel</span>, model_name <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span>
<span class="va">models</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/build_model.html" class="external-link">build_model</a></span><span class="op">(</span><span class="va">lfqp</span>, <span class="va">modelFunction</span><span class="op">)</span>

<span class="co">## compute contrasts from linear model, moderate t-statistics and p-values</span>
<span class="va">contr</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/ContrastsInterface.html" class="external-link">Contrasts</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">models</span>, <span class="va">contrasts</span><span class="op">)</span> |&gt; 
  <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/ContrastsInterface.html" class="external-link">ContrastsModerated</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span>

<span class="co"># compute contrasts using imputation imputation and moderate</span>
<span class="va">conI</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/ContrastsInterface.html" class="external-link">ContrastsSimpleImpute</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span> <span class="va">lfqp</span>, <span class="va">contrasts</span><span class="op">)</span> |&gt; 
  <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/ContrastsInterface.html" class="external-link">ContrastsModerated</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span>

<span class="co"># merge contrasts, to obtain differences for all proteins</span>
<span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/ContrastsInterface.html" class="external-link">addContrastResults</a></span><span class="op">(</span>prefer <span class="op">=</span> <span class="va">contr</span>, add <span class="op">=</span> <span class="va">conI</span><span class="op">)</span>

<span class="co">## visualize results in panels</span>
<span class="va">pl</span> <span class="op">&lt;-</span> <span class="va">contrasts</span><span class="op">$</span><span class="va">merged</span><span class="op">$</span><span class="fu">get_Plotter</span><span class="op">(</span><span class="op">)</span> 
<span class="va">panelA</span> <span class="op">&lt;-</span> <span class="va">pl</span><span class="op">$</span><span class="fu">histogram</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span>                                                        
<span class="va">panelB</span> <span class="op">&lt;-</span> <span class="va">pl</span><span class="op">$</span><span class="fu">volcano</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">FDR</span> <span class="op">+</span>
   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span>


<span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">panelA</span>, <span class="va">panelB</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="paper_files/figure-html/exampleContrasts-1.png" alt="(ref:exampleContrasts)" width="90%"><p class="caption">
(ref:exampleContrasts)
</p>
</div>
<p>R linear model and linear mixed effect models allow modeling parallel
designs, repeated measurements, factorial designs, and many more. Models
in <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> are
specified using R’s linear model and mixed model formula interface.
Therefore, knowledge of the R regression model infrastructure <span class="citation">(Faraway 2016; Venables and Ripley 2002)</span> is
advantageous when using our package. Furthermore, this glass box
approach should make it easy to reimplement an analysis performed with
<em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> using
base R or other programming languages by reading the analysis script.
However, acknowledging the R formula interface’s complexity to
non-statisticians and the popularity of <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>,
we provide the functionality to suggest a model formula from the sample
annotation provided in tabular form similarly to <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>.</p>
<p>Using the tidy table to model the data ensures interoperability with
other proteomics-related packages that manage their data with
tidy-tables, e.g., <em><a href="https://CRAN.R-project.org/package=protti" class="external-link">protti</a></em> <span class="citation">(Quast, Schuster, and Picotti 2022)</span>. The use of
R6 classes, which encapsulate the configuration and the data, allows for
writing very concise code (no function arguments needed).
Auto-completion support in the Rstudio editor makes it easy for novices
to explore <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>s functionality
(Figure @ref(fig:tabCompletion)). To simplify integration of <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> with
Bioconductor based workflows there is a method to convert the LFQData
class into a <em><a href="https://bioconductor.org/packages/3.14/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>.</p>
<p>To ease the usage barriers of the R package to users not being
familiar with statistics and R programming, we integrated some workflows
into our data management platform <a href="https://fgcz-bfabric.uzh.ch/" class="external-link">b-fabric</a> <span class="citation">(Türker et al. 2010)</span>. This integration enables
our users to select the input and basic settings in a graphical user
interface. Then, as an output, the user receives a report including
input files, the R markdown file, and R scripts necessary to replicate
the analysis using their in-house R installation. The b-fabric system
runs a computing infrastructure controlled by a local resource
management system that supports cloud-bursting <span class="citation">(Aleksiev et al. 2013)</span>. In this way, <a href="https://fgcz-bfabric.uzh.ch/" class="external-link">b-fabric</a> helps scientists to
meet requirements from funding agencies, journals, and academic
institutions to publish data according to the FAIR (Findable,
Accessible, Interoperable and Reusable) <span class="citation">(Wilkinson et al. 2016)</span> data principles.</p>
</div>
<div class="section level3">
<h3 id="benchmarking-modelling-approaches">Benchmarking modelling approaches<a class="anchor" aria-label="anchor" href="#benchmarking-modelling-approaches"></a>
</h3>
<p>The Benchmark functionality of <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> includes
receiver operator curves (ROC) and computes partial areas under those
curves (pAUC) and other scores, e.g., the false discovery proportion
FDP. We use those scores, i.e. the <span class="math inline">\(pAUC\)</span> at <span class="math inline">\(10%\)</span> FDR and the FDP, to examine how well
the methods implemented in <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> model
quantitative mass spectrometric high throughput data and compare them
with results produced by <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
and <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>.
Table @ref(tab:describeBenchmarkedModels) summarizes all modelling
methods we evaluated.</p>
<p>When comparing differential expression analysis performance, a
relevant parameter is the number of proteins for which a method
estimated differences (see Figure @ref(fig:FDRfdp) Panel A), which
indicates how robust the procedure works in the presence of missing
observations. For each protein, we tried to determine four differences
(<span class="math inline">\(\Delta = (1.20, 1.25, 1.30,
1.50)\)</span>). Given 4046 proteins with at least two peptides, there
are in total <span class="math inline">\(16184\)</span> possible
differences. However, some methods can not estimate all of them. The set
of proteins with effect size estimates might differ for each method,
biasing direct comparison of scores such as pAUC. For instance, we
observe that <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
estimates <span class="math inline">\(16058\)</span> group differences
while the mixed effect models estimates the fewest with <span class="math inline">\(15940\)</span>. Hence, to conclude that one method
shows a better performance, it does not suffice if the pAUC is greater,
but the number of proteins with differential expression results needs to
be equal or larger.</p>
<p>Figure @ref(fig:FDRfdp) Panel B shows how various estimates obtained
from the models, i.e., the difference between groups, <span class="math inline">\(t\)</span>-statistics, and scaled <span class="math inline">\(p\)</span>-values allow identifying true positives
(TP) given a false positive rate (FPR) of <span class="math inline">\(10%\)</span>, by displaying the partial area under
the ROC. Ordering proteins by the <span class="math inline">\(t\)</span>-statistic or <span class="math inline">\(p\)</span>-value leads to a higher <span class="math inline">\(pAUC_{10}\)</span> than when ordering by the
estimated difference among groups. Suppose an accurate estimate of the
difference among groups is essential. In that case, the linear models
fitted to protein intensities, calculated using Tukey’s median polish,
perform better than those directly modeling peptide intensities, e.g.,
<code>ropeca</code> or <code>prot_mixed_effect_moderated</code> (see
Table @ref(tab:describeBenchmarkedModels) Abundance column). We
speculate that outliers at the peptide level do not affect the protein
estimates when using Tukey’s median polish method, a non-parametric
method to infer protein abundances. This hypothesis could be examined,
by including other forms of protein intensity inferences implemented in
<code>prolfqua</code>, e.g. <code>top-N</code> or <code>rlm</code>, into
the benchmark.</p>
<p>There are only minor differences in the <span class="math inline">\(pAUC_{10}\)</span> between the <span class="math inline">\(t\)</span>-statistics or the scaled <span class="math inline">\(p\)</span>-value (see Figure @ref(fig:FDRfdp)
Panel A). Interestingly, the <span class="math inline">\(pAUC\)</span>
increases when using <span class="math inline">\(p\)</span>-values
instead of the <span class="math inline">\(t\)</span>-statistics for
linear models, while it decreases for mixed effect models. The reason is
an erroneous denominator degrees of freedom estimation for many proteins
in the case of the mixed effect models.</p>
<p>We also benchmark if the <span class="math inline">\(FDR\)</span>
obtained from a model is an unbiased estimate of the false discovery
proportion <span class="math inline">\(FDP\)</span>. Figure
@ref(fig:FDRfdp) Panel C draws on the horizontal axis the FDR determined
from the model, and on the vertical axis, the FDP obtained from the
confusion matrix. Most lines are below the diagonal, indicating that the
FDR estimates are modestly conservative for this benchmark dataset. In
the case of <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>,
we observe a high proportion of false discoveries for small <span class="math inline">\(FDR\)</span> values. In the case of <em><a href="https://bioconductor.org/packages/3.14/PECA" class="external-link">PECA</a></em>, the
FDR estimates, obtained by Benjamini-Hochberg correcting the regulation
probabilities, strongly overestimate the <span class="math inline">\(FDP\)</span>.</p>
<p>(ref:FDRfdp) Panel A - Number of estimated contrasts for each
modeling method (higher is better). Panel B - Partial area under the ROC
curve at <span class="math inline">\(10\%\)</span> FPR (<span class="math inline">\(pAUC_10\)</span>) for all contrasts and three
different statistics: the difference among groups, the scaled <span class="math inline">\(p\)</span>-value (sign(diff) <span class="math inline">\(\cdot\)</span> p.value) and the <span class="math inline">\(t\)</span>-statistics (higher is better). The red
line indicates the average area under the curve of all methods. Panel C
- Plots the false discovery proportion (FDP) as a function of the FDR.
Ideally, the FDR should be equal to the FDP. Therefore larger distances
from the diagonal are worse.</p>
<div class="figure">
<img src="paper_files/figure-html/FDRfdp-1.png" alt="(ref:FDRfdp)" width="90%"><p class="caption">
(ref:FDRfdp)
</p>
</div>
<p>Using a benchmark dataset whose ground truth is known (see Methods),
we assessed the performance of different modeling approaches implemented
in <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>
(Tables @ref(tab:prolfquaModels) and
@ref(tab:describeBenchmarkedModels)), <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
and <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>.
Table @ref(tab:describeBenchmarkedModels) summarizes which methods we
have evaluated, which MaxQuant results we used, and if the models are
fitted to peptide or protein intensities. We make the R-markdown files
to replicate the benchmarking available at <em><a href="https://github.com/wolski/prolfquabenchamrk" class="external-link">prolfquabenchamrk</a></em>
and at <a href="https://wolski.github.io/prolfquabenchmark" class="external-link">BenchmarkingIonstarData</a>.</p>
<p>Since only technical replicates are available for each dilution,
essential sources of variation present in every experiment, such as the
biochemical and biological, are not measured. Therefore, the dataset
captures only the variance from the chromatography, electro-spray, and
mass spectrometric measurement method. Thus, while we can extrapolate
some of the results obtained to more realistic datasets, we need to be
careful not to over-interpret the results. Specifically, the observed
variances will be higher in more realistic experiments, and the power
will be lower for the same sample sizes. Furthermore, the proportion of
missing observations in real-life datasets might also be higher.</p>
<p>We can conclude that if we want to sort the proteins according to the
likelihood of being differentially regulated to perform gene set
enrichment analysis <span class="citation">(Subramanian et al.
2005)</span>, the <span class="math inline">\(t\)</span>-statistic is
better suited than the fold-change estimate. Modeling the degrees of
freedom when computing the <span class="math inline">\(p\)</span>-values
might improve the inference. However, this improvement is minuscule (see
Figure @ref(fig:FDRfdp) panel B). There is no such improvement for the
mixed effect model, most likely because the degrees of freedom are
erroneously estimated for many models. Furthermore, for the fixed effect
linear model, the empirical Bayes variance shrinkage, as suggested by
<span class="citation">Smyth (2004)</span>, consistently improves the
ranking of proteins compared with the unmoderated estimates (not shown)
and fails to do so for mixed effect models.</p>
<p>Computing the statistics at the peptide level, e.g., the <span class="math inline">\(t\)</span>-statistics or <span class="math inline">\(p\)</span>-value, then summarizing these
statistics using their median produces the highest <span class="math inline">\(AUC\)</span> scores among all the tested models
(see Figure @ref(fig:FDRfdp) Panel A left). Furthermore, by using the
Beta distribution to model the number of peptides observed, we can
further improve the <span class="math inline">\(pAUC\)</span> scores
(see Figure @ref(fig:FDRfdp) Panel A center). However, the properties of
Beta-based probabilities are not well understood; for instance, the
p-values are not uniformly distributed under the null hypothesis (not
shown). Furthermore, the FDR estimates obtained when correcting for
multiple testing with the Benjamini-Hochberg method are biased and
overestimate the false discovery proportion (see Figure @ref(fig:FDRfdp)
Panel C). Therefore, we can not recommend this method if an unbiased
estimate of FDR is essential, which is frequently the case. In addition,
peptides are stronger affected by missing values, reducing the number of
contrasts we could estimate for the dataset using this method (see
Figure @ref(fig:FDRfdp) Panel C).</p>
<p>The probabilistic dropout analysis implemented in the <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>
produces inferences comparable to those of other methods (Figure
@ref(fig:FDRfdp) Panel A). Because of the robustness of the dropout
model to missing observations we obtain difference estimates for all
proteins and contrasts (Figure @ref(fig:FDRfdp) Panel B). Moreover, the
estimated differences are unbiased and show high diagnostic accuracy
(Figure @ref(fig:FDRfdp) Panel A). Furthermore, the performance of the
scaled <span class="math inline">\(p\)</span>-values or the <span class="math inline">\(t\)</span>-statistics is comparable with that of
the linear model with variance moderation
(<code>prot_med_lm_moderated</code>). Therefore, we are planning to
integrate the <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>
package as an additional modeling option into <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> (see Figure
@ref(fig:ContrastUML)).</p>
<p>The R-package <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em> and
<em><a href="https://bioconductor.org/packages/3.14/prolfqua" class="external-link">prolfqua</a></em>
model the missing data directly, while MSstats imputes the data using an
accelerated failure model. Despite imputation, <em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
does not estimate group differences for more proteins and does not
achieve a higher <span class="math inline">\(pAUC\)</span> score than
<em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>.
Furthermore, Figure @ref(fig:FDRfdp) Panel C shows that when using
<em><a href="https://bioconductor.org/packages/3.14/MSstats" class="external-link">MSstats</a></em>
the proportion of false discoveries might be very high even when
filtering for a low FDR.</p>
<p>We focused our benchmark on comparing the statistical modeling
methods using three different scores, while we fixed the pre-processing
steps. However, there are other equally or even more important
parameters of a protein quantification pipeline <span class="citation">(Fröhlich et al. 2022)</span>. One of them is the
normalization of the intensities within the samples to remove systematic
differences <span class="citation">(Pursiheimo et al. 2015)</span>. The
method used to infer protein intensities from peptide intensities is an
additional important factor <span class="citation">(Grossmann et al.
2010)</span>. For instance, the original <em><a href="https://bioconductor.org/packages/3.14/proDA" class="external-link">proDA</a></em>
publication uses MaxLFQ <span class="citation">(Cox et al. 2014)</span>
protein estimates. However, when using MaxLFQ intensities reported by
MaxQuant, the <span class="math inline">\(pAUC_{10}\)</span> is
significantly lower (<span class="math inline">\(pAUC_{10}(\textrm{t-statistics}) = 66%\)</span>)
compared with results obtained when protein abundances are estimated
from peptide abundances using Tukey’s median polish (<span class="math inline">\(pAUC_{10}(\textrm{t-statistics}) = 73%\)</span>).
Last but not least, the software <span class="citation">(Cox and Mann
2008; Yu et al. 2020)</span>, identifying proteins and generating the
quantification values can also significantly contribute to the
performance of the entire pipeline, altering the number of identified
proteins and the sensitivity and specificity of the differential
expression analysis.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p><em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>
allows for considerable flexibility to model quantitative proteomics
experiments. Various types of models are available (see Figure
@ref(fig:ContrastUML) and Table @ref(tab:prolfquaModels)), and the
contrast specification is explicit and consistent for all models. The
modular design of <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>, allows for
adding new features, e.g., generalized linear models (<em>glm</em>’s) to
model the presence or absence information of a protein, or robust linear
models (<em>rlm</em>’s), in the future. R’s formula interface for linear
models is flexible, widely used, and well documented <span class="citation">(Faraway 2016)</span>. We use the formula interface to
specify the models, making it easy to reproduce an analysis performed
with <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> in
other statistical methods programming languages. However, the developed
framework is flexible enough to in the future integrate other modeling
methods, e.g., the probabilistic dropout model <span class="citation">(Ahlmann-Eltze and Anders 2020)</span> or accurate
variance estimation <span class="citation">(Zhu et al. 2020)</span>.
Hence, <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em>
enables you to call various methods and makes selecting the best
differential expression analysis algorithm for your problem easy.</p>
<p>When comparing statistical modeling methods for the differential
expression analysis, we assessed performance measures such as the number
of estimated contrasts, the <span class="math inline">\(pAUC\)</span>,
and if the FDR is an unbiased estimate of the FDR. It is relevant that
an analysis pipeline shows good performance in all these categories.
Leveraging these computer experiments, we can provide the following
advice: i) estimate protein abundances from peptide abundances using a
robust or nonparametric regression method; ii) use linear models because
they show good performance in all categories; iii) if the measurements
are correlated, as for technical replicates, mixed effect models might
work if the sample sizes are large; if not, aggregate the replicates and
fit a linear model instead; iv) if you use fixed-effect linear models,
apply variance moderation to improve the <span class="math inline">\(t\)</span>-statistics and <span class="math inline">\(p\)</span>-value estimates; v) If you want to sort
your protein lists to perform gene set enrichment analysis, use the
<span class="math inline">\(t\)</span>-statistic instead of the
difference; vi) do not impute missing observation but statistically
model missingness to estimate parameters, i.e., group differences.
Finally, the differential expression analysis result obtained with
prolfqua are comparable to or better than when using other differential
expression analysis tools.</p>
<p>In summary, <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> is an
easy-to-use and feature-rich R package to analyze quantitative mass
spectrometric data with simple or complex experimental designs. It also
can generate conclusive reports and to benchmark MS software and
statistical methods. Furthermore, with minimal adaptations, this R
package can analyze different quantitative proteomics data (e.g.,
labeling-based TMT-, PRM- DIA-data). We provide documentation, in
vignette format, at the website <a href="https://github.com/fgcz/prolfqua/" class="external-link uri">https://github.com/fgcz/prolfqua/</a>. This document was
created using Rmarkdown. All the code needed to replicate the document
or the benchmark results is available at: <a href="https://github.com/wolski/prolfquabenchmark" class="external-link uri">https://github.com/wolski/prolfquabenchmark</a>.</p>
</div>
<div class="section level2">
<h2 class="unlisted unnumbered" id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>The authors thank the technology platform fund (TPF) of the
University of Zurich and all FGCZ proteomics colleagues for fruitful
discussions.</p>
</div>
<div class="section level2">
<h2 class="unlisted unnumbered" id="abbreviations">Abbreviations<a class="anchor" aria-label="anchor" href="#abbreviations"></a>
</h2>
<div style="page-break-after: always;"></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-bioxrvproDA2020" class="csl-entry">
Ahlmann-Eltze, Constantin, and Simon Anders. 2020. <span>“proDA:
Probabilistic Dropout Analysis for Identifying Differentially Abundant
Proteins in Label-Free Mass Spectrometry.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/661496" class="external-link">https://doi.org/10.1101/661496</a>.
</div>
<div id="ref-VMMAD" class="csl-entry">
Aleksiev, Tyanko, Simon Barkow-Oesterreicher, Peter Kunszt, Sergio
Maffioletti, Riccardo Murri, and Christian Panse. 2013.
<span>“<span>VM</span>-<span>MAD</span>: A Cloud/Cluster Software for
Service-Oriented Academic Environments.”</span> In <em>Lecture Notes in
Computer Science</em>, 447–61. Springer Berlin Heidelberg. <a href="https://doi.org/10.1007/978-3-642-38750-0_34" class="external-link">https://doi.org/10.1007/978-3-642-38750-0_34</a>.
</div>
<div id="ref-Bates2015FittingJSS" class="csl-entry">
Bates, Douglas, Martin Mächler, Ben Bolker, and Steve Walker. 2015.
<span>“Fitting Linear Mixed-Effects Models Using Lme4.”</span>
<em>Journal of Statistical Software, Articles</em> 67 (1): 1–48. <a href="https://doi.org/10.18637/jss.v067.i01" class="external-link">https://doi.org/10.18637/jss.v067.i01</a>.
</div>
<div id="ref-bolstad2003comparison" class="csl-entry">
Bolstad, Benjamin M, Rafael A Irizarry, Magnus Åstrand, and Terence P.
Speed. 2003. <span>“A Comparison of Normalization Methods for High
Density Oligonucleotide Array Data Based on Variance and Bias.”</span>
<em>Bioinformatics</em> 19 (2): 185–93.
</div>
<div id="ref-bruderer2015extending" class="csl-entry">
Bruderer, Roland, Oliver M Bernhardt, Tejas Gandhi, Saša M Miladinović,
Lin-Yang Cheng, Simon Messner, Tobias Ehrenberger, et al. 2015.
<span>“Extending the Limits of Quantitative Proteome Profiling with
Data-Independent Acquisition and Application to Acetaminophen-Treated
Three-Dimensional Liver Microtissues.”</span> <em>Molecular &amp;
Cellular Proteomics</em> 14 (5): 1400–1410.
</div>
<div id="ref-Bubis2017" class="csl-entry">
Bubis, Julia A., Lev I. Levitsky, Mark V. Ivanov, Irina A. Tarasova, and
Mikhail V. Gorshkov. 2017. <span>“Comparative Evaluation of Label-Free
Quantification Methods for Shotgun Proteomics.”</span> <em>Rapid
Communications in Mass Spectrometry</em> 31 (7): 606–12. <a href="https://doi.org/10.1002/rcm.7829" class="external-link">https://doi.org/10.1002/rcm.7829</a>.
</div>
<div id="ref-POMA2021" class="csl-entry">
Castellano-Escuder, Pol, Cristina Andrés-Lacueva, and Alex Sánchez-Pla.
2021. <em>POMA: User-Friendly Workflow for Metabolomics and Proteomics
Data Analysis</em>. <a href="https://github.com/pcastellanoescuder/POMA" class="external-link">https://github.com/pcastellanoescuder/POMA</a>.
</div>
<div id="ref-R6cite" class="csl-entry">
Chang, Winston. 2020. <em>R6: Encapsulated Classes with Reference
Semantics</em>. <a href="https://CRAN.R-project.org/package=R6" class="external-link">https://CRAN.R-project.org/package=R6</a>.
</div>
<div id="ref-Choi2014" class="csl-entry">
Choi, Meena, Ching-Yun Chang, Timothy Clough, Daniel Broudy, Trevor
Killeen, Brendan MacLean, and Olga Vitek. 2014. <span>“MSstats: An
<span>R</span> Package for Statistical Analysis of Quantitative Mass
Spectrometry-Based Proteomic Experiments.”</span>
<em>Bioinformatics</em> 30 (17): 2524–26.
</div>
<div id="ref-cox2014accurate" class="csl-entry">
Cox, Jürgen, Marco Y Hein, Christian A Luber, Igor Paron, Nagarjuna
Nagaraj, and Matthias Mann. 2014. <span>“Accurate Proteome-Wide
Label-Free Quantification by Delayed Normalization and Maximal Peptide
Ratio Extraction, Termed MaxLFQ.”</span> <em>Molecular &amp; Cellular
Proteomics</em> 13 (9): 2513–26.
</div>
<div id="ref-Cox2008MaxQuant" class="csl-entry">
Cox, Jürgen, and Matthias Mann. 2008. <span>“MaxQuant Enables High
Peptide Identification Rates, Individualized p.p.b.-Range Mass
Accuracies and Proteome-Wide Protein Quantification.”</span> <em>Nature
Biotechnology</em> 26 (December): 1367–72. <a href="https://doi.org/10.1038/nbt.1511" class="external-link">https://doi.org/10.1038/nbt.1511</a>.
</div>
<div id="ref-demichev2020dia" class="csl-entry">
Demichev, Vadim, Christoph B Messner, Spyros I Vernardis, Kathryn S
Lilley, and Markus Ralser. 2020. <span>“DIA-NN: Neural Networks and
Interference Correction Enable Deep Proteome Coverage in High
Throughput.”</span> <em>Nature Methods</em> 17 (1): 41–44.
</div>
<div id="ref-eng2015deeper" class="csl-entry">
Eng, Jimmy K, Michael R Hoopmann, Tahmina A Jahan, Jarrett D Egertson,
William S Noble, and Michael J MacCoss. 2015. <span>“A Deeper Look into
Comet—Implementation and Features.”</span> <em>Journal of the American
Society for Mass Spectrometry</em> 26 (11): 1865–74.
</div>
<div id="ref-faraway2016extending" class="csl-entry">
Faraway, Julian J. 2016. <em>Extending the Linear Model with
<span>R</span>: Generalized Linear, Mixed Effects and Nonparametric
Regression Models</em>. Chapman; Hall/CRC.
</div>
<div id="ref-frohlich2022benchmarking" class="csl-entry">
Fröhlich, Klemens, Eva Brombacher, Matthias Fahrner, Daniel Vogele,
Lucas Kook, Niko Pinter, Peter Bronsert, et al. 2022.
<span>“Benchmarking of Analysis Strategies for Data-Independent
Acquisition Proteomics Using a Large-Scale Dataset Comprising
Inter-Patient Heterogeneity.”</span> <em>Nature Communications</em> 13
(1): 1–13.
</div>
<div id="ref-Goeminne2016" class="csl-entry">
Goeminne, Ludger JE, Kris Gevaert, and Lieven Clement. 2016.
<span>“Peptide-Level Robust Ridge Regression Improves Estimation,
Sensitivity, and Specificity in Data-Dependent Quantitative Label-Free
Shotgun Proteomics.”</span> <em>Molecular &amp; Cellular Proteomics</em>
15 (2): 657–68.
</div>
<div id="ref-goeminne2020msqrob" class="csl-entry">
Goeminne, Ludger JE, Adriaan Sticker, Lennart Martens, Kris Gevaert, and
Lieven Clement. 2020. <span>“MSqRob Takes the Missing Hurdle: Uniting
Intensity-and Count-Based Proteomics.”</span> <em>Analytical
Chemistry</em> 92 (9): 6278–87.
</div>
<div id="ref-Grossmann2010" class="csl-entry">
Grossmann, Jonas, Bernd Roschitzki, Christian Panse, Claudia Fortes,
Simon Barkow-Oesterreicher, Dorothea Rutishauser, and Ralph Schlapbach.
2010. <span>“Implementation and Evaluation of Relative and Absolute
Quantification in Shotgun Proteomics with Label-Free Methods.”</span>
<em>Journal of Proteomics</em> 73 (9): 1740–46. <a href="https://doi.org/10.1016/j.jprot.2010.05.011" class="external-link">https://doi.org/10.1016/j.jprot.2010.05.011</a>.
</div>
<div id="ref-huber2002variance" class="csl-entry">
Huber, Wolfgang, Anja Von Heydebreck, Holger Sültmann, Annemarie
Poustka, and Martin Vingron. 2002. <span>“Variance Stabilization Applied
to Microarray Data Calibration and to the Quantification of Differential
Expression.”</span> <em>Bioinformatics</em> 18 (suppl_1): S96–104.
</div>
<div id="ref-ph525xseries" class="csl-entry">
Irizarry, Rafael, and Michael Love. 2018. <span>“PH525x Series -
Biomedical Data Science.”</span> 2018. <a href="http://genomicsclass.github.io/book/pages/interactions_and_contrasts.html" class="external-link">http://genomicsclass.github.io/book/pages/interactions_and_contrasts.html</a>.
</div>
<div id="ref-kong2017msfragger" class="csl-entry">
Kong, Andy T, Felipe V Leprevost, Dmitry M Avtonomov, Dattatreya
Mellacheruvu, and Alexey I Nesvizhskii. 2017. <span>“MSFragger:
Ultrafast and Comprehensive Peptide Identification in Mass
Spectrometry–Based Proteomics.”</span> <em>Nature Methods</em> 14 (5):
513–20.
</div>
<div id="ref-caret2008" class="csl-entry">
Kuhn, Max. 2008. <span>“Building Predictive Models in <span>R</span>
Using the Caret Package.”</span> <em>Journal of Statistical Software,
Articles</em> 28 (5): 1–26. <a href="https://doi.org/10.18637/jss.v028.i05" class="external-link">https://doi.org/10.18637/jss.v028.i05</a>.
</div>
<div id="ref-Kuznetsova2017lmerTest" class="csl-entry">
Kuznetsova, Alexandra, Per Brockhoff, and Rune Christensen. 2017.
<span>“lmerTest Package: Tests in Linear Mixed Effects Models.”</span>
<em>Journal of Statistical Software, Articles</em> 82 (13): 1–26. <a href="https://doi.org/10.18637/jss.v082.i13" class="external-link">https://doi.org/10.18637/jss.v082.i13</a>.
</div>
<div id="ref-laubscher2021baf" class="csl-entry">
Laubscher, Dominik, Berkley E Gryder, Benjamin D Sunkel, Thorkell
Andresson, Marco Wachtel, Sudipto Das, Bernd Roschitzki, et al. 2021.
<span>“BAF Complexes Drive Proliferation and Block Myogenic
Differentiation in Fusion-Positive Rhabdomyosarcoma.”</span> <em>Nature
Communications</em> 12 (1): 1–16.
</div>
<div id="ref-law2020guide" class="csl-entry">
Law, Charity W, Kathleen Zeglinski, Xueyi Dong, Monther Alhamdoosh,
Gordon K Smyth, and Matthew E Ritchie. 2020. <span>“A Guide to Creating
Design Matrices for Gene Expression Experiments.”</span>
<em>F1000Research</em> 9.
</div>
<div id="ref-de2022apoe2" class="csl-entry">
Leeuw, Sherida M de, Aron WT Kirschner, Karina Lindner, Ruslan Rust,
Vanessa Budny, Witold E Wolski, Anne-Claude Gavin, Roger M Nitsch, and
Christian Tackenberg. 2022. <span>“Apoe2, E3, and E4 Differentially
Modulate Cellular Homeostasis, Cholesterol Metabolism, and Inflammatory
Response in Isogenic iPSC-Derived Astrocytes.”</span> <em>Stem Cell
Reports</em> 17 (1): 110–26.
</div>
<div id="ref-maclean2010skyline" class="csl-entry">
MacLean, Brendan, Daniela M Tomazela, Nicholas Shulman, Matthew
Chambers, Gregory L Finney, Barbara Frewen, Randall Kern, David L Tabb,
Daniel C Liebler, and Michael J MacCoss. 2010. <span>“Skyline: An Open
Source Document Editor for Creating and Analyzing Targeted Proteomics
Experiments.”</span> <em>Bioinformatics</em> 26 (7): 966–68.
</div>
<div id="ref-mcgurk2020use" class="csl-entry">
McGurk, Kathryn A, Arianna Dagliati, Davide Chiasserini, Dave Lee,
Darren Plant, Ivona Baricevic-Jones, Janet Kelsall, et al. 2020.
<span>“The Use of Missing Values in Proteomic Data-Independent
Acquisition Mass Spectrometry to Enable Disease Activity
Discrimination.”</span> <em>Bioinformatics</em> 36 (7): 2217–23.
</div>
<div id="ref-meier2021reduced" class="csl-entry">
Meier-Abt, Fabienne, Witold E Wolski, Ge Tan, Sandra Kummer, Sabine
Amon, Markus G Manz, Ruedi Aebersold, and Alexandre Theocharides. 2021.
<span>“Reduced Cxcl4/Pf4 Expression as a Driver of Increased Human
Hematopoietic Stem and Progenitor Cell Proliferation in Polycythemia
Vera.”</span> <em>Blood Cancer Journal</em> 11 (2): 1–6.
</div>
<div id="ref-pursiheimo2015optimization" class="csl-entry">
Pursiheimo, Anna, Anni P Vehmas, Saira Afzal, Tomi Suomi, Thaman Chand,
Leena Strauss, Matti Poutanen, Anne Rokka, Garry L Corthals, and Laura L
Elo. 2015. <span>“Optimization of Statistical Methods Impact on
Quantitative Proteomics Data.”</span> <em>Journal of Proteome
Research</em> 14 (10): 4118–26.
</div>
<div id="ref-quast2022protti" class="csl-entry">
Quast, Jan-Philipp, Dina Schuster, and Paola Picotti. 2022.
<span>“Protti: An r Package for Comprehensive Data Analysis of
Peptide-and Protein-Centric Bottom-up Proteomics Data.”</span>
<em>Bioinformatics Advances</em> 2 (1): vbab041.
</div>
<div id="ref-RCoreTeam2021" class="csl-entry">
R Core Team. 2021. <span>“R: A Language and Environment for Statistical
Computing.”</span> Vienna, Austria: R Foundation for Statistical
Computing. 2021. <a href="https://www.R-project.org/" class="external-link">https://www.R-project.org/</a>.
</div>
<div id="ref-Ritchie2015" class="csl-entry">
Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law,
Wei Shi, and Gordon K Smyth. 2015. <span>“Limma Powers Differential
Expression Analyses for RNA-Sequencing and Microarray Studies.”</span>
<em>Nucleic Acids Research</em> 43 (7): e47–47.
</div>
<div id="ref-rost2014openswath" class="csl-entry">
Röst, Hannes L, George Rosenberger, Pedro Navarro, Ludovic Gillet, Saša
M Miladinović, Olga T Schubert, Witold Wolski, et al. 2014.
<span>“OpenSWATH Enables Automated, Targeted Analysis of
Data-Independent Acquisition MS Data.”</span> <em>Nature
Biotechnology</em> 32 (3): 219–23.
</div>
<div id="ref-RStudio2022" class="csl-entry">
RStudio Team. 2022. <em>RStudio: Integrated Development Environment for
r</em>. Boston, MA: RStudio, Inc. <a href="http://www.rstudio.com/" class="external-link">http://www.rstudio.com/</a>.
</div>
<div id="ref-rubin1976inference" class="csl-entry">
Rubin, Donald B. 1976. <span>“Inference and Missing Data.”</span>
<em>Biometrika</em> 63 (3): 581–92.
</div>
<div id="ref-shen2018ionstar" class="csl-entry">
Shen, Xiaomeng, Shichen Shen, Jun Li, Qiang Hu, Lei Nie, Chengjian Tu,
Xue Wang, et al. 2018. <span>“IonStar Enables High-Precision,
Low-Missing-Data Proteomics Quantification in Large Biological
Cohorts.”</span> <em>Proceedings of the National Academy of
Sciences</em> 115 (21): E4767–76.
</div>
<div id="ref-Smyth2004linear" class="csl-entry">
Smyth, Gordon K. 2004. <span>“Linear Models and Empirical Bayes Methods
for Assessing Differential Expression in Microarray Experiments.”</span>
<em>Statistical Applications in Genetics and Molecular Biology</em> 3
(1): 1–25.
</div>
<div id="ref-subramanian2005gene" class="csl-entry">
Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee,
Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005.
<span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for
Interpreting Genome-Wide Expression Profiles.”</span> <em>Proceedings of
the National Academy of Sciences</em> 102 (43): 15545–50.
</div>
<div id="ref-Suomi2017bEnhanced" class="csl-entry">
Suomi, Tomi, and Laura L Elo. 2017. <span>“Enhanced Differential
Expression Statistics for Data-Independent Acquisition
Proteomics.”</span> <em>Scientific Reports</em> 7: 5869.
</div>
<div id="ref-tan2022proteomic" class="csl-entry">
Tan, Ge, Witold E Wolski, Sandra Kummer, Mara Hofstetter, Alexandre PA
Theocharides, Markus G Manz, Ruedi Aebersold, and Fabienne Meier-Abt.
2022. <span>“Proteomic Identification of Proliferation and Progression
Markers in Human Polycythemia Vera Stem and Progenitor Cells.”</span>
<em>Blood Advances</em>.
</div>
<div id="ref-teo2014saintexpress" class="csl-entry">
Teo, Guoci, Guomin Liu, Jianping Zhang, Alexey I Nesvizhskii,
Anne-Claude Gingras, and Hyungwon Choi. 2014. <span>“SAINTexpress:
Improvements and Additional Features in Significance Analysis of
INTeractome Software.”</span> <em>Journal of Proteomics</em> 100: 37–43.
</div>
<div id="ref-tukey1977exploratory" class="csl-entry">
Tukey, John W et al. 1977. <em>Exploratory Data Analysis</em>. Vol. 2.
Reading, MA.
</div>
<div id="ref-bfabric" class="csl-entry">
Türker, Can, Fuat Akal, Dieter Joho, Christian Panse, Simon
Barkow-Oesterreicher, Hubert Rehrauer, and Ralph Schlapbach. 2010.
<span>“B-Fabric: The Swiss Army Knife for Life Sciences.”</span> In
<em>Proceedings of the 13th International Conference on Extending
Database Technology - <span>EDBT</span> <span></span>10</em>.
<span>ACM</span> Press. <a href="https://doi.org/10.1145/1739041.1739135" class="external-link">https://doi.org/10.1145/1739041.1739135</a>.
</div>
<div id="ref-valikangas2016systematic" class="csl-entry">
Välikangas, Tommi, Tomi Suomi, and Laura L Elo. 2016. <span>“A
Systematic Evaluation of Normalization Methods in Quantitative
Label-Free Proteomics.”</span> <em>Briefings in Bioinformatics</em> 19
(1): 1–11.
</div>
<div id="ref-da2020philosopher" class="csl-entry">
Veiga Leprevost, Felipe da, Sarah E Haynes, Dmitry M Avtonomov, Hui-Yin
Chang, Avinash K Shanmugam, Dattatreya Mellacheruvu, Andy T Kong, and
Alexey I Nesvizhskii. 2020. <span>“Philosopher: A Versatile Toolkit for
Shotgun Proteomics Data Analysis.”</span> <em>Nature Methods</em> 17
(9): 869–70.
</div>
<div id="ref-MASS2002" class="csl-entry">
Venables, W. N., and B. D. Ripley. 2002. <em>Modern Applied Statistics
with <span>S</span></em>. Fourth. New York: Springer. <a href="https://www.stats.ox.ac.uk/pub/MASS4/" class="external-link">https://www.stats.ox.ac.uk/pub/MASS4/</a>.
</div>
<div id="ref-vidova2017review" class="csl-entry">
Vidova, Veronika, and Zdenek Spacil. 2017. <span>“A Review on Mass
Spectrometry-Based Quantitative Proteomics: Targeted and Data
Independent Acquisition.”</span> <em>Analytica Chimica Acta</em> 964:
7–23.
</div>
<div id="ref-tidydata2014" class="csl-entry">
Wickham, Hadley. 2014. <span>“Tidy Data.”</span> <em>The Journal of
Statistical Software</em> 59. <a href="https://doi.org/10.18637/jss.v059.i10" class="external-link">https://doi.org/10.18637/jss.v059.i10</a>.
</div>
<div id="ref-tidyverse2019" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy
D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019.
<span>“Welcome to the <span class="nocase">tidyverse</span>.”</span>
<em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686" class="external-link">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-FAIR" class="csl-entry">
Wilkinson, Mark D., Michel Dumontier, IJsbrand Jan Aalbersberg,
Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al.
2016. <span>“The <span>FAIR</span> Guiding Principles for Scientific
Data Management and Stewardship.”</span> <em>Scientific Data</em> 3 (1).
<a href="https://doi.org/10.1038/sdata.2016.18" class="external-link">https://doi.org/10.1038/sdata.2016.18</a>.
</div>
<div id="ref-prolfquadata" class="csl-entry">
Wolski, Witold. 2021. <span>“Prolfquadata.”</span> 2021. <a href="https://gitlab.bfabric.org/wolski/prolfquadata" class="external-link">https://gitlab.bfabric.org/wolski/prolfquadata</a>.
</div>
<div id="ref-yu2020fast" class="csl-entry">
Yu, Fengchao, Sarah E Haynes, Guo Ci Teo, Dmitry M Avtonomov, Daniel A
Polasky, and Alexey I Nesvizhskii. 2020. <span>“Fast Quantitative
Analysis of timsTOF PASEF Data with MSFragger and IonQuant.”</span>
<em>BioRxiv</em>.
</div>
<div id="ref-yu2016pipi" class="csl-entry">
Yu, Fengchao, Ning Li, and Weichuan Yu. 2016. <span>“PIPI: PTM-Invariant
Peptide Identification Using Coding Method.”</span> <em>Journal of
Proteome Research</em> 15 (12): 4423–35.
</div>
<div id="ref-DEP2018" class="csl-entry">
Zhang, Xiaofei, Arne H Smits, Gabrielle BA van Tilburg, Huib Ovaa,
Wolfgang Huber, and Michiel Vermeulen. 2018. <span>“Proteome-Wide
Identification of Ubiquitin Interactions Using UbIA-MS.”</span>
<em>Nature Protocols</em> 13: 530–50.
</div>
<div id="ref-zhu2020deqms" class="csl-entry">
Zhu, Yafeng, Lukas M Orre, Yan Zhou Tran, Georgios Mermelekas, Henrik J
Johansson, Alina Malyutina, Simon Anders, and Janne Lehtiö. 2020.
<span>“DEqMS: A Method for Accurate Variance Estimation in Differential
Protein Expression Analysis.”</span> <em>Molecular &amp; Cellular
Proteomics</em> 19 (6): 1047–57.
</div>
</div>
<div style="page-break-after: always;"></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 class="unlisted unnumbered" id="creating-a-prolfqua-configuration">Creating a prolfqua configuration<a class="anchor" aria-label="anchor" href="#creating-a-prolfqua-configuration"></a>
</h3>
<p>The following code demonstrates how we use <em><a href="https://github.com/fgcz/prolfqua" class="external-link">prolfqua</a></em> to analyze
protein intensities reported in the MSFragger
<code>combined_protein.tsv</code> file. First, we create a tidy table
containing the protein abundances by reading the
<code>combined_protein.tsv</code> file using
<code>tidy_MSFragger_combined_protein.</code> Then, we read the sample
annotation from the file <code>annotation.xlsx</code> file. Next, we
create an <code>AnalysisTableAnnotation</code> R6 object. Bottom-up
proteomics data is hierarchical, i.e., a protein has peptides, peptides
might be modified, etc. Therefore, the
<code>AnalysisTableAnnotation</code> has a <code>hierarchy</code> field
storing a list with an entry for each hierarchy level. Since
<code>combined_portein.tsv</code> only holds protein level data, the
hierarchy list has one element, and we use it to specify which column
contains the protein identifiers. We also need to define which column
contains the protein abundances we want to use for the data analysis.
Finally, we have to specify which columns contain the explanatory
variables of the analysis. The <code>AnalysisTableAnnotation</code> has
the field <code>factors,</code> a list with as many entries as
explanatory variables. Here we include two explanatory variables, the
dilution, specified in the column ‘sample’, and ‘run’ stored in the
column ‘run_ID’, representing the order of the measurement.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">datadir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">find.package</a></span><span class="op">(</span><span class="st">"prolfquadata"</span><span class="op">)</span> , <span class="st">"quantdata"</span><span class="op">)</span>
<span class="va">inputFragfile</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"MSFragger_IonStar2018_PXD003881.zip"</span><span class="op">)</span>
<span class="va">inputAnnotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"annotation_Ionstar2018_PXD003881.xlsx"</span><span class="op">)</span>
<span class="co"># read input annotation</span>
<span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span><span class="op">(</span><span class="va">inputAnnotation</span><span class="op">)</span>

<span class="va">protein</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">unz</a></span><span class="op">(</span><span class="va">inputFragfile</span>,<span class="st">"IonstarWithMSFragger/combined_protein.tsv"</span><span class="op">)</span>,
             header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span>

<span class="co"># read combined_protein.tsv </span>
<span class="va">protein</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/MSFragger.html" class="external-link">tidy_MSFragger_combined_protein</a></span><span class="op">(</span><span class="va">protein</span><span class="op">)</span>
<span class="co"># remove proteins identified by a single peptide</span>
<span class="va">protein</span> <span class="op">&lt;-</span> <span class="va">protein</span> |&gt; <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">unique.stripped.peptides</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span>

<span class="co"># annotate the data</span>
<span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">annotation</span>, <span class="va">protein</span><span class="op">)</span>
<span class="va">atable</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/AnalysisTableAnnotation.html" class="external-link">AnalysisTableAnnotation</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span>
<span class="va">atable</span><span class="op">$</span><span class="va">fileName</span> <span class="op">=</span> <span class="st">"raw.file"</span>
<span class="co"># specify column containing protein identifiers</span>
<span class="va">atable</span><span class="op">$</span><span class="va">hierarchy</span><span class="op">[[</span><span class="st">"protein_Id"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"protein"</span>

<span class="co"># column with protein abundances</span>
<span class="va">atable</span><span class="op">$</span><span class="fu">setWorkIntensity</span><span class="op">(</span><span class="st">"total.intensity"</span><span class="op">)</span>

<span class="co"># the factors of the analysis</span>
<span class="va">atable</span><span class="op">$</span><span class="va">factors</span><span class="op">[[</span><span class="st">"dilution."</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"sample"</span>
<span class="va">atable</span><span class="op">$</span><span class="va">factors</span><span class="op">[[</span><span class="st">"run"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="st">"run_ID"</span>

<span class="va">config</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/AnalysisConfiguration.html" class="external-link">AnalysisConfiguration</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">atable</span><span class="op">)</span>

<span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prolfqua/man/setup_analysis.html" class="external-link">setup_analysis</a></span><span class="op">(</span><span class="va">merged</span>, <span class="va">config</span><span class="op">)</span>
<span class="va">lfqdata</span> <span class="op">&lt;-</span> <span class="fu">prolfqua</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/prolfqua/man/LFQData.html" class="external-link">LFQData</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">adata</span>, <span class="va">config</span><span class="op">)</span>
<span class="co"># remove small intensities</span>
<span class="va">lfqdata</span><span class="op">$</span><span class="fu">remove_small_intensities</span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 class="unlisted unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span>
<span class="co">## Running under: macOS Monterey 12.4</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRblas.0.dylib</span>
<span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [7] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] lmerTest_3.1-3 lme4_1.1-29    Matrix_1.4-0  </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] nlme_3.1-155        fs_1.5.2            webshot_0.5.3      </span>
<span class="co">##   [4] RColorBrewer_1.1-3  progress_1.2.2      httr_1.4.3         </span>
<span class="co">##   [7] rprojroot_2.0.3     prolfqua_0.4.1      numDeriv_2016.8-1.1</span>
<span class="co">##  [10] tools_4.1.2         backports_1.4.1     bslib_0.3.1        </span>
<span class="co">##  [13] utf8_1.2.2          R6_2.5.1            vipor_0.4.5        </span>
<span class="co">##  [16] lazyeval_0.2.2      colorspace_2.0-3    tidyselect_1.1.2   </span>
<span class="co">##  [19] gridExtra_2.3       prettyunits_1.1.1   compiler_4.1.2     </span>
<span class="co">##  [22] rvest_1.0.2         textshaping_0.3.6   cli_3.3.0          </span>
<span class="co">##  [25] xml2_1.3.3          desc_1.4.1          plotly_4.10.0      </span>
<span class="co">##  [28] labeling_0.4.2      bookdown_0.26       sass_0.4.1         </span>
<span class="co">##  [31] scales_1.2.0        pkgdown_2.0.3       systemfonts_1.0.4  </span>
<span class="co">##  [34] stringr_1.4.0       digest_0.6.29       minqa_1.2.4        </span>
<span class="co">##  [37] svglite_2.1.0       rmarkdown_2.14      pkgconfig_2.0.3    </span>
<span class="co">##  [40] htmltools_0.5.2     fastmap_1.1.0       highr_0.9          </span>
<span class="co">##  [43] readxl_1.4.0        htmlwidgets_1.5.4   rlang_1.0.2        </span>
<span class="co">##  [46] rstudioapi_0.13     farver_2.1.0        jquerylib_0.1.4    </span>
<span class="co">##  [49] generics_0.1.2      jsonlite_1.8.0      dplyr_1.0.9        </span>
<span class="co">##  [52] car_3.0-13          magrittr_2.0.3      kableExtra_1.3.4   </span>
<span class="co">##  [55] Rcpp_1.0.8.3        ggbeeswarm_0.6.0    munsell_0.5.0      </span>
<span class="co">##  [58] fansi_1.0.3         abind_1.4-5         lifecycle_1.0.1    </span>
<span class="co">##  [61] stringi_1.7.6       yaml_2.3.5          carData_3.0-5      </span>
<span class="co">##  [64] MASS_7.3-55         grid_4.1.2          ggrepel_0.9.1      </span>
<span class="co">##  [67] forcats_0.5.1       crayon_1.5.1        lattice_0.20-45    </span>
<span class="co">##  [70] cowplot_1.1.1       splines_4.1.2       hms_1.1.1          </span>
<span class="co">##  [73] knitr_1.39          pillar_1.7.0        ggpubr_0.4.0       </span>
<span class="co">##  [76] boot_1.3-28         ggsignif_0.6.3      glue_1.6.2         </span>
<span class="co">##  [79] evaluate_0.15       data.table_1.14.2   BiocManager_1.30.18</span>
<span class="co">##  [82] vctrs_0.4.1         nloptr_2.0.3        cellranger_1.1.0   </span>
<span class="co">##  [85] gtable_0.3.0        purrr_0.3.4         tidyr_1.2.0        </span>
<span class="co">##  [88] cachem_1.0.6        ggplot2_3.3.6       conflicted_1.1.0   </span>
<span class="co">##  [91] xfun_0.31           broom_0.8.0         rstatix_0.7.0      </span>
<span class="co">##  [94] ragg_1.2.2          viridisLite_0.4.0   tibble_3.1.7       </span>
<span class="co">##  [97] pheatmap_1.0.12     beeswarm_0.4.0      memoise_2.0.1      </span>
<span class="co">## [100] ellipsis_0.3.2      BiocStyle_2.22.0</span></code></pre>
</div>
<div class="section level3">
<h3 class="unlisted unnumbered" id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a>
</h3>
<p>(ref:tabCompletion) The screenshot displays the command-line
completion (tab completion) of RStudio on the
<code><a href="https://rdrr.io/pkg/prolfqua/man/LFQData.html" class="external-link">prolfqua::LFQData</a></code> R6 object. In the example, it shows the
getter methods of the object.</p>
<div class="figure">
<img src="codeSnippet1TabCompletion.png" alt="(ref:tabCompletion)" width="66%"><p class="caption">
(ref:tabCompletion)
</p>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"prolfqua"</span><span class="op">)</span>, 
          <span class="st">"Figures/hexStickerProlfqua.png"</span><span class="op">)</span> |&gt;
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="../../../r-site-library_prolfqua/prolfqua/Figures/hexStickerProlfqua.png" alt="Sticker maintainer: Witold E. Wolski; License: Creative Commons Attribution CC-BY. Feel free to share and adapt, but don't forget to credit the author." height="5cm"><p class="caption">
Sticker maintainer: Witold E. Wolski; License: Creative Commons
Attribution CC-BY. Feel free to share and adapt, but don’t forget to
credit the author.
</p>
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="https://fgcz.ch" class="external-link">Functional Genomics Center
Zurich (FGCZ)</a> - <a href="https://uzh.ch" class="external-link">University of Zurich</a>/<a href="https://ethz.ch" class="external-link">ETH Zurich</a>, Winterthurerstrasse 190, CH-8057
Zurich, Switzerland. <span class="math inline">\(^\ddagger\)</span><a href="https://www.sib.swiss/" class="external-link">Swiss Institute of Bioinformatics (SIB),
Quartier Sorge - Batiment Amphipole, 1015 Lausanne, Switzerland</a>
<span class="math inline">\(^\dagger\)</span>Correspondence: <a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
